### PR TITLE
[Xamarin.Android.Build.Tasks] Provide file.encoding for dx tool

### PR DESCRIPTION
**Pull request by dellis1972**
Fixes: http://work.devdiv.io/784551
Fixes: https://developercommunity.visualstudio.com/content/problem/437807/v-isual-studio-2019-preview-20.html
Fixes: #2685

Context: #2688 (comment)
Context: #2688 (comment)

In commit 8571a7f we updated the `<CompileToDalvik/>` task to
generate and use a "response file" via the `--input-list` argument.
This helped reduce the overall length of the command line arguments
passed.  It had a side effect of causing a problem if the path
contains non-ASCII characters, causing the following error on Windows:

	java.io.FileNotFoundException: C:\Temp\Scratch.├£ml├ñ├╝ts\Scratch.├£ml├ñ├╝ts\obj\Debug\90\android\bin\classes.zip (The system cannot find the path specified)

This is because while the `input-list` file itself is fine, the paths
within it might contain non-ASCII characters.  We need to let java
know that, which we can do that via the `java` argument:

	-Dfile.encoding=UTF8

The `<CompileToDalvik/>` task can be run in two modes, the first is
just invoking the `dx` executable, and the second is invoking
`dx.jar` via `java`.  Both of these need to be supported.

For the normal `java` invocation we can just pass the argument with
`-Dfile.encoding=UTF8`.  However for `dx` that is not a valid
argument.  We instead need to provide it via:

	-JDfile.encoding=UTF8

The `-J` tells `dx` that this is a `java` argument.  `dx` will then
take everything after the `-J` , prepend a `-` and pass that to
`java`.

Finally, there was a [problem of Unicode Normalization Forms][0]
and `string.Replace()`.  On macOS, it is possible within the
`<AndroidComputeResPaths/>` task for `$(ProjectDir)` to use a
different filesystem normalization than `@(AndroidAsset)`, because
`@(AndroidAsset)` comes from a directory traversal and will use
Unicode Normalization Form D [(-ish)][1] on HFS+, while
`$(ProjectDir)` will come from the `msbuild` command-line invocation
performed by the unit tests and will be in Unicode Normalization
Form C, and through "luck" `string.Replace()` doesn't handle this:

	var e_acute = "\u00e9";          // http://www.fileformat.info/info/unicode/char/e9/index.htm
	var e_formc = e_acute.Normalize (System.Text.NormalizationForm.FormC);
	var e_formd = e_acute.Normalize (System.Text.NormalizationForm.FormD);
	e_formc.StartsWith (e_formd);    // true
	e_formc.Replace (e_formd, "");   // == "é" == e_formc; no replacement performed!

`<AndroidComputeResPaths>` uses `string.Replace()` to convert a "full"
path name into a "relative" path name, but as the `string.Replace()`
didn't do anything here with the mismatched Unicode Normalization,
`<AndroidComputeResPaths/>` would return *lowercased full paths* to
subsequent tasks, not the expected lowercased *relative* paths, which
ultimately results in a `Resource.designer.cs` which is *empty*:

	Task "AndroidComputeResPaths" (TaskId:62)
	    ...
	    IntermediateFiles: (TaskId:62)
	      /users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android/bin/testrelease/temp/buildreleaseappwitha initandümläüts(truedx)/resources/drawable-hdpi/icon.png (TaskId:62)
	      ... wat?! ...
	...
	MainActivity.cs(22,20,22,28): error CS0103: The name 'Resource' does not exist in the current context

Fix this problem by normalizing the strings into the same Unicode
Normalization Form *before* performing the `string.Replace()` call,
which works as expected/intended, restoring the semantics of
`<AndroidComputeResPaths>`.

Update one of the unit tests has been to use non-ASCII characters.

[0]: https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android-pr-builder-release/623/testReport/junit/Xamarin.Android.Build.Tests.BuildTest/BuildApplicationWithSpacesInPathAndÜmläüts/_True__dx________Release/
[1]: https://en.wikipedia.org/wiki/HFS_Plus#Design

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs [hunk 1]
```diff
@@ -113,19 +120,19 @@ protected override string GenerateCommandLineCommands ()
 					foreach (var l in AlternativeJarFiles) {
 						var fullPath = Path.GetFullPath (l.ItemSpec);
 						Log.LogDebugMessage ($"    {fullPath}");
-						sw.WriteLine (fullPath);
+						sw.WriteLine (fullPath.Normalize (NormalizationForm.FormC));
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs]: What normalization form, if any, does NTFS/Fat32 use to store filenames?

Apparently, [*no* normalization form is specified](https://docs.microsoft.com/en-us/windows/desktop/FileIO/naming-a-file):

> There is no need to perform any Unicode normalization on path and file name strings for use by the Windows file I/O API functions because the file system treats path and file names as an opaque sequence of **WCHAR**s.

`AlternativeJarFiles` is coming from MSBuild/the file system, and thus *presumably* will be "correct" for file I/O purposes, in turn suggesting that we *shouldn't* normalize this filename, as such normalization may *prevent* the file from being found via e.g. `File.Open()` or whatever Java uses.

**Comment by dellis1972** [comment]
Seems `aapt2` has a problem with special characters too

```
Executing compile -o obj\Release\flata\compiled.flata --dir "E:\A\_work\484\s\bin\TestDebug\temp\BuildReleaseAppWithA InItAndÜmläüts(Falsed8r8)\obj\Release\res" (TaskId:99)
2019-02-03T22:24:21.9981627Z   E:\A\_work\484\s\bin\TestDebug\temp\BuildReleaseAppWithA InItAnd├£ml├ñ├╝ts(Falsed8r8)\obj\Release\res : error APT0000: failed to open directory: The system cannot find the file specified. (2). [E:\A\_work\484\s\bin\TestDebug\temp\BuildReleaseAppWithA InItAndÜmläüts(Falsed8r8)\UnnamedProject.csproj]
2019-02-03T22:24:21.9981877Z   
```

**Comment by dellis1972** [comment]
Ok looks like it works under `aapt`

```
Task "CompileToDalvik" (TaskId:187)
  Task Parameter:DxJarPath=C:\Users\dean\android-toolchain\sdk\build-tools\28.0.3\\lib\dx.jar (TaskId:187)
  Task Parameter:DxExtraArguments=--dex --no-strict (TaskId:187)
  Task Parameter:JavaToolPath=C:\Program Files\Java\jdk1.8.0_131\bin (TaskId:187)
  Task Parameter:JavaMaximumHeapSize=1G (TaskId:187)
  Task Parameter:ClassesOutputDirectory=obj\Release\android\bin\classes (TaskId:187)
  Task Parameter:MultiDexEnabled=False (TaskId:187)
  Task Parameter:MultiDexMainDexListFile=obj\Release\multidex.keep (TaskId:187)
  Task Parameter:
      JavaLibrariesToCompile=
          c:\Sandbox\xamarin-android\bin\Debug\lib\xamarin.android\xbuild-frameworks\MonoAndroid\v9.0\mono.android.jar
          Hello (World).jar (TaskId:187)
    processing ClassesOutputDirectory... (TaskId:187)
      c:\Sandbox\xamarin-android\bin\TestDebug\temp\BuildReleaseAppWithA InItAndÜmläüts(Falsed8)\obj\Release\android\bin\classes.zip (TaskId:187)
      c:\Sandbox\xamarin-android\bin\Debug\lib\xamarin.android\xbuild-frameworks\MonoAndroid\v9.0\mono.android.jar (TaskId:187)
      c:\Sandbox\xamarin-android\bin\TestDebug\temp\BuildReleaseAppWithA InItAndÜmläüts(Falsed8)\Hello (World).jar (TaskId:187)
  C:\Program Files\Java\jdk1.8.0_131\bin\java.exe -Dfile.encoding=UTF8 -Xmx1G -jar C:\Users\dean\android-toolchain\sdk\build-tools\28.0.3\\lib\dx.jar --dex --no-strict --
  input-list=C:\Users\dean\AppData\Local\Temp\tmp44F6.tmp --output obj\Release\android\bin  (TaskId:187)
Done executing task "CompileToDalvik". (TaskId:187)
``` 

It seems `aapt2` has a problem with non ascii characters. So we won't be able to confirm this works on master since we have `aapt2` enabled by default.

**Comment by dellis1972** [comment]
Yay, fixed windows... broke Mac :(

**Comment by jonpryor** [comment]
Actually, I think Wikipedia may be old; *my* APFS *does* appear to perform some normalization.

Fun app!

```csharp
using System;
using System.IO;
using System.Text;

class App {
	public static void Main (string[] args)
	{
		var c = Encoding.UTF8.GetString (new byte[]{(byte) 0xc3, (byte)0xa9});
		var d = Encoding.UTF8.GetString (new byte[]{(byte) 0x65, (byte)0xcc, (byte)0x81});
		foreach (var dir in args) {
			var dc = Path.Combine (dir, c + ".txt");
			var dd = Path.Combine (dir, d + ".txt");

			Console.WriteLine ($"# Creating FormC: {dc}");
			using (var o = File.CreateText (dc)) {
				o.WriteLine ("FormC");
			}

			Console.WriteLine ($"# Creating FormD: {dd}");
			using (var o = File.CreateText (dd)) {
				o.WriteLine ("FormD");
			}

			Console.WriteLine ($"# Reading: {dc}");
			Console.WriteLine (File.ReadAllText (dc));
			Console.WriteLine ($"# Reading: {dd}");
			Console.WriteLine (File.ReadAllText (dd));
			
			
			foreach (var f in Directory.EnumerateFiles (dir, c + ".txt")) {
				Console.WriteLine ($"# EnumerateFiles({c}) (FormC)");
				Console.WriteLine (File.ReadAllText (f));
			}
			foreach (var f in Directory.EnumerateFiles (dir, d + ".txt")) {
				Console.WriteLine ($"# EnumerateFiles({d}) (FormD)");
				Console.WriteLine (File.ReadAllText (f));
			}
		}
	}
}
```

This tries to create the files `é.txt` and `é.txt` within the directories provided on the command-line, writes text to them, then tries to read them.

On macOS, they're the same file (but see note below!); `/tmp` is APFS:

```bash
$ mono encoding.exe  /tmp
# Creating FormC: /tmp/é.txt
# Creating FormD: /tmp/é.txt
# Reading: /tmp/é.txt
FormD

# Reading: /tmp/é.txt
FormD

# EnumerateFiles(é) (FormC)
FormD
```

Note that while we *write and read* a filename containing Normalization Form C, we read "FormD", meaning we're reading the contents of the *second  file created*, which happens to be the file with Normalization Form D.  (If we create the FormD variant before the FormC variant, we read FormC content; the *last* `File.CreateText()` "wins".)

This is somewhat akin to case insensitivity: opening `a.txt` is the same as opening `A.txt`.

Worrying is that the `Directory.EnumerateFiles(dir, d + ".txt")` invocation found *nothing*.

If we run the app on an HFS+ volume (external spinning-disk drive):

```
$ mono encoding.exe  /Volumes/Seagate4TB
# Creating FormC: /Volumes/Seagate4TB/é.txt
# Creating FormD: /Volumes/Seagate4TB/é.txt
# Reading: /Volumes/Seagate4TB/é.txt
FormD

# Reading: /Volumes/Seagate4TB/é.txt
FormD

# EnumerateFiles(é) (FormD)
FormD
```

As with APFS, we're writing to the "same file" *twice*, so the last `File.CreateText()` "wins", hence "both" files containing the text `FormD`.

Notice that on HFS+ `Directory.EnumerateFiles(dir, d + ".txt")` finds a file, while `Directory.EnumerateFiles(dir, c + ".txt")` *does **not***; it's the opposite behavior as APFS.

On macOS (Mojave, 10.14.2), *reading and writing* to files is normalized: Whether you open a file using normalization Form C or Form D, it doesn't matter.

What *isn't* normalized is *directory traversal*, i.e. **opendir**(3)/`Directory.EnumerateFiles()` (which ultimately should boil down to **opendir**(3)...).  *Here* we can see the difference between APFS and HFS+: while both normalize filenames, APFS appears to normalize to FormC, while HFS+ normalizes to FormD.  (Not that I can readily find documentation verifying this, but this is the behavior I'm observing.)

---

Compare to Windows:

```
>.\encoding.exe %TMP%
# Creating FormC: …\AppData\Local\Temp\é.txt
# Creating FormD: …\AppData\Local\Temp\e'.txt
# Reading: …\AppData\Local\Temp\é.txt
FormC

# Reading: …\AppData\Local\Temp\e'.txt
FormD

# EnumerateFiles(é) (FormC)
FormC

# EnumerateFiles(e') (FormD)
FormD
```

Note here that the FormC and FormD files are treated entirely *separately*, as if it were a case-sensitive filesystem (`a` isn't the same file as `A`). Further note that I ran this in `CMD.EXE`, and it doesn't render FormD properly.

(PowerShell ISE fares *worse* in rendering `é`, which I found quite surprising.)

Viewing `%TMP%` within Explorer is also interesting, as there are "two" files named `é.txt`.

**Comment by jonpryor** [comment]
[Previous macOS investigation](https://github.com/xamarin/xamarin-android/pull/2688#issuecomment-460887811) was inadequate; What I *should* have done was:

```csharp
			Console.WriteLine ($"# Creating FormC: {dc}");
			using (var o = File.CreateText (dc)) {
				o.WriteLine ("FormC");
			}
			Console.ReadLine();
```

Then during the `Console.ReadLine()` look at the `ls -lFtr $DIR | hexdump -C` output to see the encoding of what was *actually* created.

Alternative, I should have written to *new* empty directories, not `/tmp`, and nuked that directory after every run, so let's go with that.

Consider two variants of `encoding.cs`: one which writes the FormC file first, and one which writes the FormD file first (swap lines 14-17 with lines 19-22). Call them `encoding-c.exe` and `encoding-d.exe`.

### APFS

```
$ rm -Rf a; mkdir a; mono encoding-c.exe  a
# Creating FormC: a/é.txt
# Creating FormD: a/é.txt
# Reading: a/é.txt
FormD

# Reading: a/é.txt
FormD

# EnumerateFiles(é) (FormC)
FormD

$ rm -Rf a; mkdir a; mono encoding-d.exe  a
# Creating FormD: a/é.txt
# Creating FormC: a/é.txt
# Reading: a/é.txt
FormC

# Reading: a/é.txt
FormC

# EnumerateFiles(é) (FormD)
FormC
```

With APFS and a clean directory, the encoding of the created file is the encoding of the *first* created file.  Directory enumeration in turn uses the normalization of the first created file, *even if* that file was opened via the "second" filename.  Thus, `encoding-c.exe` prints `EnumerateFiles(é) (FormC)`, while `encoding-d.exe` prints `EnumerateFiles(é) (FormD)`.


### HFS+

```
$ rm -Rf /Volumes/Seagate4TB/a; mkdir /Volumes/Seagate4TB/a ; mono encoding-c.exe /Volumes/Seagate4TB/a
# Creating FormC: /Volumes/Seagate4TB/a/é.txt
# Creating FormD: /Volumes/Seagate4TB/a/é.txt
# Reading: /Volumes/Seagate4TB/a/é.txt
FormD

# Reading: /Volumes/Seagate4TB/a/é.txt
FormD

# EnumerateFiles(é) (FormD)
FormD

$ rm -Rf /Volumes/Seagate4TB/a; mkdir /Volumes/Seagate4TB/a ; mono encoding-d.exe /Volumes/Seagate4TB/a
# Creating FormD: /Volumes/Seagate4TB/a/é.txt
# Creating FormC: /Volumes/Seagate4TB/a/é.txt
# Reading: /Volumes/Seagate4TB/a/é.txt
FormC

# Reading: /Volumes/Seagate4TB/a/é.txt
FormC

# EnumerateFiles(é) (FormD)
FormC
```

With HFS+, the filesystem is *always* Normalization Form D (-ish).

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt2.cs [hunk 2]
```diff
@@ -34,7 +34,7 @@ public class Aapt2 : AsyncTask {
 
 		protected string ResourceDirectoryFullPath (string resourceDirectory)
 		{
-			return (Path.IsPathRooted (resourceDirectory) ? resourceDirectory : Path.Combine (WorkingDirectory, resourceDirectory)).TrimEnd ('\\');
+			return (Path.IsPathRooted (resourceDirectory) ? resourceDirectory : resourceDirectory).TrimEnd ('\\');
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/Aapt2.cs]: Doesn't this line thus become `resourceDirectory.TrimEnd('\\')`?

File: src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs [hunk 3]
```diff
@@ -120,12 +127,12 @@ protected override string GenerateCommandLineCommands ()
...
 					if (File.Exists (zip)) {
 						Log.LogDebugMessage ($"    {zip}");
-						sw.WriteLine (Path.GetFullPath (zip));
+						sw.WriteLine (Path.GetFullPath (zip).Normalize (NormalizationForm.FormC));
 					}
 					foreach (var jar in JavaLibrariesToCompile) {
 						var fullPath = Path.GetFullPath (jar.ItemSpec);
 						Log.LogDebugMessage ($"    {fullPath}");
-						sw.WriteLine (fullPath);
+						sw.WriteLine (fullPath.Normalize (NormalizationForm.FormC));
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs]: Assuming my prior musings are remotely accurate, this change should *not* be necessary.  `JavaLibariesToCompile` comes from filesystem reads (`<DetermineJavaLibrariesToCompile/>` and co), and thus should be assumed to have the correct filesystem encoding/(non-)normalization.

File: src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs [hunk 4]
```diff
@@ -120,12 +127,12 @@ protected override string GenerateCommandLineCommands ()
 					var zip = Path.GetFullPath (Path.Combine (ClassesOutputDirectory, "..", "classes.zip"));
 					if (File.Exists (zip)) {
 						Log.LogDebugMessage ($"    {zip}");
-						sw.WriteLine (Path.GetFullPath (zip));
+						sw.WriteLine (Path.GetFullPath (zip).Normalize (NormalizationForm.FormC));
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs]: Likewise, `ClassesOutputDirectory` is "from MSBuild" ("hardcoded" in the `.targets`, but `$(IntermediateOutputPath)` is plausibly overridable by the user and thus could contain non-ASCII characters), and as such we should *not* further normalize this value.

File: src/Xamarin.Android.Build.Tasks/Tasks/AndroidUpdateResDir.cs [hunk 5]
```diff
@@ -96,8 +97,8 @@ public override bool Execute ()
 					if (string.IsNullOrEmpty (rel)) {
 						rel = item.GetMetadata ("Identity");
 						if (!string.IsNullOrEmpty (ProjectDir)) {
-							var fullRelPath = Path.GetFullPath (rel);
-							var fullProjectPath = Path.GetFullPath (ProjectDir);
+							var fullRelPath = Path.GetFullPath (rel).Normalize (NormalizationForm.FormC);
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/AndroidUpdateResDir.cs]: The normalization here *is* required; the result of this normalization isn't used for subsequent filesystem reads/written to a file, but is instead used for relative path computation, and thus should be fine.  I think.

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt2.cs (hunk 2 above)
Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Tasks/Aapt2.cs]: Now that I think about it. we can't do this.
The problem is the MSBuild may well change the WorkingDirectory is a Node is reused, this can happen when we use Yield(). So if we remove this code , we will get an error eventually as the working directory will not contain the non-rooted paths. 
This was the reason why we use the `WorkingDirectory` property in the first place. We store it when the task starts so we know where we are supposed to be. 
Now what we might be able to do is set the working directory of the `aapt2` tool when we run it. That might work

Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Tasks/Aapt2.cs]: Turns out we already run `aapt2` in the `WorkingDirectory` so we should be ok... in theory.

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt2.cs [hunk 6]
```diff
@@ -34,7 +34,7 @@ public class Aapt2 : AsyncTask {
 
 		protected string ResourceDirectoryFullPath (string resourceDirectory)
 		{
-			return (Path.IsPathRooted (resourceDirectory) ? resourceDirectory : Path.Combine (WorkingDirectory, resourceDirectory)).TrimEnd ('\\');
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/Aapt2.cs]: I still don't understand why this change is required.  The method name *implies* that a full path should be returned, which explains the original `Path.IsPathRooted()` and `Path.Combine(WorkingDirectory, ...)` calls, but now...  Presumably a relative path could be returned, but the method name still says "FullPath".

What's going on here?

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/Builder.cs [hunk 7]
```diff
@@ -347,37 +347,40 @@ protected bool BuildInternal (string projectOrSolution, string target, string []
 			var start = DateTime.UtcNow;
 			var args  = new StringBuilder ();
 			var psi   = new ProcessStartInfo (XABuildExe);
+			var responseFile = Path.Combine (XABuildPaths.TestOutputDirectory, Path.GetDirectoryName (projectOrSolution), "msbuild.rsp");
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/Builder.cs]: How is this *used*?  Or is `msbuild.rsp` "special" and MSBuild automatically reads it?

The [docs](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-response-files?view=vs-2017#msbuildrsp) are not entirely clear:

> This file, *MSBuild.rsp*, must be in the same directory as *MSBuild.exe*, otherwise it will not be found. 

We almost certainly are not writing `msbuild.rsp` into the same directory as `MSBuild.exe`...

Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/Builder.cs]: That is different from the docs from `msbuild /?`

```
@<file>            Insert command-line settings from a text file. To specify
                     multiple response files, specify each response file
                     separately.

                     Any response files named "msbuild.rsp" are automatically
                     consumed from the following locations:
                     (1) the directory of msbuild.exe
                     (2) the directory of the first project or solution built
```

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/ManagedResourceParserTests.cs [hunk 8]
```diff
@@ -74,9 +74,9 @@ public class ManagedResourceParserTests : BaseTest {
 ";
 
 		[Test]
-		public void GenerateDesignerFile ()
+		public void GenerateDesignerFileWithÜmläüts ()
 		{
-			var path = Path.Combine ("temp", TestName);
+			var path = Path.Combine ("temp", TestName + " Some Space");
```
[the remark this answers is not in the document]
Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/ManagedResourceParserTests.cs]: The `TestName` already contains the ü.

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt2Compile.cs [hunk 9]
```diff
@@ -94,7 +94,7 @@ protected string GenerateCommandLineCommands (ITaskItem dir, string outputArchiv
 			cmd.AppendSwitchIfNotNull ("-o ", outputArchive);
 			if (!string.IsNullOrEmpty (ResourceSymbolsTextFile))
 				cmd.AppendSwitchIfNotNull ("--output-text-symbols ", ResourceSymbolsTextFile);
-			cmd.AppendSwitchIfNotNull ("--dir ", ResourceDirectoryFullPath (dir.ItemSpec));
+			cmd.AppendSwitchIfNotNull ("--dir ", dir.ItemSpec.TrimEnd ("\\"));
```
Comment (dellis1972) [inline src/Xamarin.Android.Build.Tasks/Tasks/Aapt2Compile.cs]: Turns out `ResourceDirectoryFullPath` is used to fix up the error messages for VS. So lets try to not full path the directory and just `TrimEnd`.

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/Builder.cs (hunk 7 above)
Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/Builder.cs]: Its also different from the behaviour I have seen from actual MSBuild.
